[SPARK-26837][SQL] Pruning nested fields from object serializers

viirya opened the pull request:
## What changes were proposed in this pull request?

In SPARK-26619, we make change to prune unnecessary individual serializers when serializing objects. This is extension to SPARK-26619. We can further prune nested fields from object serializers if they are not used.

For example, in following query, we only use one field in a struct column:

```scala
val data = Seq((("a", 1), 1), (("b", 2), 2), (("c", 3), 3))
val df = data.toDS().map(t => (t._1, t._2 + 1)).select("_1._1")
```

So, instead of having a serializer to create a two fields struct, we can prune unnecessary field from it. This is what this PR proposes to do.

In order to make this change conservative and safer, a SQL config is added to control it. It is disabled by default.

TODO: Support to prune nested fields inside MapType's key and value.

## How was this patch tested?

Added tests.

viirya committed:
Moved schema pruning classes.

viirya committed:
Pruning nested fields in object serializers.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7730/
Test PASSed.

SparkQA commented:
**[Test build #102063 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102063/testReport)** for PR 23740 at commit [`e23b72e`](https://github.com/apache/spark/commit/e23b72e7965badc3accbeb65a07f42c429c69eae).

SparkQA commented:
**[Test build #102063 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102063/testReport)** for PR 23740 at commit [`e23b72e`](https://github.com/apache/spark/commit/e23b72e7965badc3accbeb65a07f42c429c69eae).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class RootField(field: StructField, derivedFromAtt: Boolean,`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102063/
Test PASSed.

viirya commented:
cc @dbtsai you might be interested in this, and @cloud-fan, @dongjoon-hyun

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -561,10 +561,31 @@ object ColumnPruning extends Rule[LogicalPlan] {
+
+      if (SQLConf.get.serializerNestedSchemaPruningEnabled && fields.nonEmpty) {
+        // Prunes nested fields in serialzers.
+        val prunedSchema = SchemaPruning.pruneDataSchema(
```
I think we can skip pruning if `prunedSchema` is as same as original one. But I keep it simpler now for review. Can update it later.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:
```diff
@@ -0,0 +1,212 @@
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types._
+
+private[sql] object SchemaPruning {
```
Except for `collectStructType` and `pruneSerializer`, other methods are moved here from `ParquetSchemaPruning`.

dongjoon-hyun replied:
As @HyukjinKwon suggested, let's remove `private[sql]` here, too.

felixcheung commented:
cool!

HyukjinKwon commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ProjectionOverSchema.scala:
```diff
@@ -26,7 +25,7 @@ import org.apache.spark.sql.types._
  * class is motivated by columnar nested schema pruning.
  */
-private[execution] case class ProjectionOverSchema(schema: StructType) {
+private[sql] case class ProjectionOverSchema(schema: StructType) {
```
Since `catalyst` is already an internal module,  `private[sql]` would better be removed.  These were removed out as of [SPARK-16813](https://issues.apache.org/jira/browse/SPARK-16813).

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruning.scala:
```diff
@@ -33,6 +32,8 @@ import org.apache.spark.sql.types.{ArrayType, DataType, MapType, StructField, St
  * SQL column, and a nested Parquet column corresponds to a [[StructField]].
  */
 private[sql] object ParquetSchemaPruning extends Rule[LogicalPlan] {
```
Ditto for execution module as well as of [SPARK-16964](https://issues.apache.org/jira/browse/SPARK-16964).

dongjoon-hyun commented:
Thank you for pinging me, @viirya !

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:
```diff
@@ -0,0 +1,212 @@
+   * right, recursively. That is, left is a "subschema" of right, ignoring order of
+   * fields.
+   */
+  def sortLeftFieldsByRight(left: DataType, right: DataType): DataType =
```
`def` -> `private def`.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:
```diff
@@ -0,0 +1,212 @@
+   * Returns the set of fields from the Parquet file that the query plan needs.
+   */
+  def identifyRootFields(projects: Seq[NamedExpression],
+      filters: Seq[Expression]): Seq[RootField] = {
```
Shall we indent like the following?
```scala
def identifyRootFields(
    projects: Seq[NamedExpression],
    filters: Seq[Expression]): Seq[RootField] = {
```

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:
```diff
@@ -0,0 +1,212 @@
+          sortLeftFieldsByRight(leftElementType, rightElementType),
+          containsNull)
+      case (MapType(leftKeyType, leftValueType, containsNull),
+      MapType(rightKeyType, rightValueType, _)) =>
```
indentation?

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:
```diff
@@ -0,0 +1,212 @@
+   * When expr is an [[Attribute]], construct a field around it and indicate that that
+   * field was derived from an attribute.
+   */
+  def getRootFields(expr: Expression): Seq[RootField] = {
```
`def` -> `private def`

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruningSuite.scala:
```diff
@@ -0,0 +1,56 @@
+          StructField("a", StructType.fromDDL("a int, b int")) ::
+            StructField("b", IntegerType) :: Nil),
+        StructType.fromDDL("a int, b int"))
+
```
Remove the empty line?

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetStructFieldObject.scala:
```diff
@@ -15,17 +15,16 @@
  * ordinal instead of the field itself.
  */
-private[execution] object GetStructFieldObject {
+private[sql] object GetStructFieldObject {
```
ditto. Remove `private[sql]`.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:
```diff
@@ -0,0 +1,212 @@
+  /**
+   * Collects all struct types from given data type object, recursively. Supports struct and array
+   * types for now.
+   * TODO: support map type.
```
Just curious. Is it difficult for map type?

viirya replied:
I'd say it's not difficult. But it needs more than the pruning logic for struct and array of struct. To keep this PR not too much big (now it is relatively large), I leave it as a TODO.

viirya replied:
I created SPARK-26847 to track it. Once this gets in, I will make a PR to finish it.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -561,10 +561,31 @@ object ColumnPruning extends Rule[LogicalPlan] {
       val prunedSerializer = s.serializer.filter(usedRefs.contains)
-      p.copy(child = SerializeFromObject(prunedSerializer, s.child))
+
+      val fields = SchemaPruning.identifyRootFields(p.projectList, Seq.empty)
```
Maybe, `rootFields` is more clearer than `fields`?

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -1680,6 +1681,62 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
+        case s: SerializeFromObject => s
+      }.head
+
+      val structs = serializer.serializer(0).collect {
```
`serializer(0)` -> `serializer.head`.

viirya replied:
I originally wanted to keep the test simple. But I think we should also test other serializers too. I updated the test case.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -1680,6 +1681,62 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
     checkAnswer(ds, Seq(Row("a"), Row("b"), Row("c")))
   }
 
+  test("SPARK-26837:Prune nested serializers") {
```
space: `SPARK-26837:Prune` -> `SPARK-26837: Prune`.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -1680,6 +1681,62 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
+      testSerializer(df4, Seq(Seq("_2", "_3"), Seq("_1")))
+      checkAnswer(df4, Seq(Row("aa", 1), Row("bb", 2), Row("cc", 3)))
+
+      val arrayData = Seq((Seq(("a", 1, ("a_1", 11)), ("b", 2, ("b_1", 22))), 1, ("aa", 1.0)),
```
Shall we split this into another test case because this function is a little long?

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:
```diff
@@ -0,0 +1,212 @@
+   * types for now.
+   * TODO: support map type.
+   */
+  def collectStructType(dt: DataType, structs: ArrayBuffer[StructType]): ArrayBuffer[StructType] = {
```
Is this public for only testing purpose at `SchemaPruningSuite`?

viirya replied:
Yes.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:
```diff
@@ -0,0 +1,212 @@
+
+  /**
+   * Filters the schema from the given file by the requested fields.
+   * Schema field ordering from the file is preserved.
```
Since this is in `catalyst` module from now, we had better update this description more properly.

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruningSuite.scala:
```diff
@@ -0,0 +1,56 @@
+      val structs = SchemaPruning.collectStructType(dt, ArrayBuffer.empty[StructType])
+      assert(structs === expected(idx))
+    }
+  }
```
Since `pruneDataSchema` becomes public, could you add a test coverage like the following here?
```scala
test("prune data schema") {
  ...
}
```

dongjoon-hyun replied:
Thank you for adding the test coverage.

dongjoon-hyun commented:
BTW, @viirya . Let's remove `[SPARK-26837]` from the test case names. We usually don't add `[SPARK-XXXXX]` for new feature implementation. It's used for BUG issue.

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruningSuite.scala:
```diff
@@ -0,0 +1,56 @@
+class SchemaPruningSuite extends SparkFunSuite {
+
+  test("collect struct types") {
+    val datatypes = Seq(
```
`datatypes` -> `dataTypes`.

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruningSuite.scala:
```diff
@@ -0,0 +1,56 @@
+          StructField("b", IntegerType) :: Nil)
+    )
+
+    val expected = Seq(
```
`expected` -> `expectedTypes`?

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruningSuite.scala:
```diff
@@ -0,0 +1,56 @@
+      ArrayType(StructType.fromDDL("a int, b int, c string")),
+      StructType(
+        StructField("a", StructType.fromDDL("a int, b int")) ::
+          StructField("b", IntegerType) :: Nil)
```
```scala
-      StructType(
-        StructField("a", StructType.fromDDL("a int, b int")) ::
-          StructField("b", IntegerType) :: Nil)
+      StructType.fromDDL("a struct<a:int, b:int>, b int")
```

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruningSuite.scala:
```diff
@@ -0,0 +1,56 @@
+      Seq(
+        StructType(
+          StructField("a", StructType.fromDDL("a int, b int")) ::
+            StructField("b", IntegerType) :: Nil),
```
ditto.
```scala
-        StructType(
-          StructField("a", StructType.fromDDL("a int, b int")) ::
-            StructField("b", IntegerType) :: Nil),
+        StructType.fromDDL("a struct<a:int, b:int>, b int"),
```

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:
```diff
@@ -0,0 +1,212 @@
+    }
+
+  /**
+   * Returns the set of fields from the Parquet file that the query plan needs.
```
Please remove `from the Parquet file` and update this properly because this is irrelevant to `Parquet` from now.

viirya committed:
Address comments from reviewers.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7746/
Test PASSed.

SparkQA commented:
**[Test build #102086 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102086/testReport)** for PR 23740 at commit [`d40fdcf`](https://github.com/apache/spark/commit/d40fdcf98be3d7076fda0105603322980e0176e9).

viirya committed:
Split and enhance test case.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7747/
Test PASSed.

SparkQA commented:
**[Test build #102087 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102087/testReport)** for PR 23740 at commit [`8136803`](https://github.com/apache/spark/commit/8136803c6d06fc7f23aca8533799385c96642c89).

viirya commented:
Thanks @felixcheung @HyukjinKwon @dongjoon-hyun! I think I have addressed your comments now except for https://github.com/apache/spark/pull/23740#discussion_r254578211 that asks to add some tests for `pruneDataSchema`. I will add some tests for it today or tomorrow.

viirya committed:
Remove another one private[sql].

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7748/
Test PASSed.

SparkQA commented:
**[Test build #102088 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102088/testReport)** for PR 23740 at commit [`5de4845`](https://github.com/apache/spark/commit/5de4845ac08a40a13fb61ddfb4ae160e6ea056c3).

SparkQA commented:
**[Test build #102088 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102088/testReport)** for PR 23740 at commit [`5de4845`](https://github.com/apache/spark/commit/5de4845ac08a40a13fb61ddfb4ae160e6ea056c3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102088/
Test FAILed.

viirya commented:
retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7749/
Test PASSed.

SparkQA commented:
**[Test build #102089 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102089/testReport)** for PR 23740 at commit [`5de4845`](https://github.com/apache/spark/commit/5de4845ac08a40a13fb61ddfb4ae160e6ea056c3).

SparkQA commented:
**[Test build #102086 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102086/testReport)** for PR 23740 at commit [`d40fdcf`](https://github.com/apache/spark/commit/d40fdcf98be3d7076fda0105603322980e0176e9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ProjectionOverSchema(schema: StructType) `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102086/
Test PASSed.

SparkQA commented:
**[Test build #102087 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102087/testReport)** for PR 23740 at commit [`8136803`](https://github.com/apache/spark/commit/8136803c6d06fc7f23aca8533799385c96642c89).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102087/
Test PASSed.

gatorsmile commented:
@hvanhovell You are the best person to review this.

SparkQA commented:
**[Test build #102089 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102089/testReport)** for PR 23740 at commit [`5de4845`](https://github.com/apache/spark/commit/5de4845ac08a40a13fb61ddfb4ae160e6ea056c3).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102089/
Test FAILed.

viirya committed:
Add test for pruneDataSchema.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7751/
Test PASSed.

SparkQA commented:
**[Test build #102092 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102092/testReport)** for PR 23740 at commit [`6fc8096`](https://github.com/apache/spark/commit/6fc8096dd7e22e29a07cb7d11ef1b96907fcc132).

dongjoon-hyun commented:
Thank you for updating, @viirya .

SparkQA commented:
**[Test build #102092 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102092/testReport)** for PR 23740 at commit [`6fc8096`](https://github.com/apache/spark/commit/6fc8096dd7e22e29a07cb7d11ef1b96907fcc132).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102092/
Test PASSed.

viirya commented:
Thanks @gatorsmile. @hvanhovell Can you take a look if you are available?

viirya commented:
ping @cloud-fan Can you take a look when you have time? Thanks.

viirya commented:
ping...

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:
```diff
@@ -0,0 +1,214 @@
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types._
+
+object SchemaPruning {
```
can you point out which methods are new in this file?

viirya replied:
`collectStructType` and `pruneSerializer` are new methods.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:
```diff
@@ -0,0 +1,214 @@
+   * prune nested fields in struct and array of struct.
+   * TODO(SPARK-26847): support to prune nested fields in key and value of map type.
+   */
+  def pruneSerializer(
```
this is specific to the column pruning of object serializers, shall we move this method there instead of putting it in the util?

viirya replied:
OK. I will do it later.

viirya committed:
Address comment.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala:
```diff
@@ -93,6 +95,72 @@ case class SerializeFromObject(
   override def output: Seq[Attribute] = serializer.map(_.toAttribute)
 }
 
+object SerializeFromObject {
```
@cloud-fan Not sure if this is proper location. If not, please let me know. Thanks.

cloud-fan commented:
shall we create a new rule to do column pruning for object serializers? And in the `ColumnPrunning` rule we skip `Project(SerializeObject(...))`

viirya commented:
> shall we create a new rule to do column pruning for object serializers? And in the `ColumnPrunning` rule we skip `Project(SerializeObject(...))`

I'd say it might be a good idea, because the column pruning grows too long. I will do it in next update.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8074/
Test PASSed.

SparkQA commented:
**[Test build #102505 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102505/testReport)** for PR 23740 at commit [`6360af2`](https://github.com/apache/spark/commit/6360af23f4eae4ff3324ea56d08c5b02676b92b8).

SparkQA commented:
**[Test build #102505 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102505/testReport)** for PR 23740 at commit [`6360af2`](https://github.com/apache/spark/commit/6360af23f4eae4ff3324ea56d08c5b02676b92b8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

viirya committed:
Move object serializer pruning to a new rule.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8116/
Test PASSed.

SparkQA commented:
**[Test build #102555 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102555/testReport)** for PR 23740 at commit [`7b33f23`](https://github.com/apache/spark/commit/7b33f23f4d14507ab6966cf201e4cec734384bcc).

SparkQA commented:
**[Test build #102555 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102555/testReport)** for PR 23740 at commit [`7b33f23`](https://github.com/apache/spark/commit/7b33f23f4d14507ab6966cf201e4cec734384bcc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102555/
Test FAILed.

dongjoon-hyun commented:
Retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8118/
Test PASSed.

SparkQA commented:
**[Test build #102557 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102557/testReport)** for PR 23740 at commit [`7b33f23`](https://github.com/apache/spark/commit/7b33f23f4d14507ab6966cf201e4cec734384bcc).

SparkQA commented:
**[Test build #102557 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102557/testReport)** for PR 23740 at commit [`7b33f23`](https://github.com/apache/spark/commit/7b33f23f4d14507ab6966cf201e4cec734384bcc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102557/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8126/
Test PASSed.

SparkQA commented:
**[Test build #102570 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102570/testReport)** for PR 23740 at commit [`49b650f`](https://github.com/apache/spark/commit/49b650f64073376c3c762139b2db27a1570a024a).

SparkQA commented:
**[Test build #102570 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102570/testReport)** for PR 23740 at commit [`49b650f`](https://github.com/apache/spark/commit/49b650f64073376c3c762139b2db27a1570a024a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102570/
Test PASSed.

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ObjectSerializerPruningSuite.scala:
```diff
@@ -0,0 +1,78 @@
+      }.asInstanceOf[NamedExpression]
+
+      // `name` in `GetStructField` affects `comparePlans`. Maybe we can ignore
+      // `name` in `GetStructField.equals`?
```
So, is this test case removing `name` just for `comparePlans`?

viirya replied:
Yes, @dongjoon-hyun.

viirya commented:
@cloud-fan Are there anymore comments? Thanks.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala:
```diff
@@ -93,6 +95,72 @@ case class SerializeFromObject(
+   * types for now.
+   * TODO(SPARK-26847): support map type.
+   */
+  def collectStructType(dt: DataType, structs: ArrayBuffer[StructType]): ArrayBuffer[StructType] = {
```
shall we move these 2 methods to `ObjectSerializerPruning`?

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -1680,6 +1681,76 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
+    }
+  }
+
+  test("Prune nested serializers: struct") {
```
Shall we create a new `DatasetOptimizationSuite` and put these new tests there?

cloud-fan commented:
looks good except a few minor comments

viirya commented:
Thanks @cloud-fan! I will address the comments tomorrow.

viirya committed:
Move methods and tests.

viirya committed:
Merge remote-tracking branch 'upstream/master' into nested-pruning-serializer-2

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8306/
Test PASSed.

SparkQA commented:
**[Test build #102800 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102800/testReport)** for PR 23740 at commit [`ef39272`](https://github.com/apache/spark/commit/ef392723a3ed8bc53332e15fc98c34ca4ff92cfb).

SparkQA commented:
**[Test build #102800 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102800/testReport)** for PR 23740 at commit [`ef39272`](https://github.com/apache/spark/commit/ef392723a3ed8bc53332e15fc98c34ca4ff92cfb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102800/
Test PASSed.

cloud-fan commented:
thanks, merging to master!
